lib/spack/spack/version.py

```diff
@@ -468,9 +449,128 @@ def intersection(self, other):
         else:
             return VersionList()
 
+
+class GitVersion(VersionBase):
+    def __init__(self, string):
+        string = str(string)  # In case we got a VersionBase or GitVersion object
+        self.is_ref = False
+
+        if string.startswith('git.'):
+            string = string[4:]
+            self.is_ref = True
+
+        self.is_commit = len(string) == 40 and COMMIT_VERSION.match(string)
```

Comment:
it looks like this allows commit versions to start with `git.`, but `Version()` doesn't.  Need a test to ensure that both work.

Reply:
In other version types, `string` preserves the *exact* original string, but here we lose `git.`.  I actually think it makes sense to keep that... shouldn't we?

Reply:
IMO it makes sense to maintain the exact string the user specified for the spec in the version. This will make it clearer when reviewing a list of specs from `spack find`

Reply:
`Version()` already does allow `git.<commit hash>`. The test I've added confirms that.

The reason I have `self.string` strip `git.` out and `str(self)` include it is that we often need the git ref alone. I can change that to be some other attribute and keep the full string in `self.string` if you prefer.

Reply:
Yeah I think it's better to be consistent with other versions and keep the original input string for output.